File: huffman.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <string>
#include <vector>
#include <tuple>

using namespace std;

// Huffman Tree Node
struct Node {
    char ch;
    int freq;
    Node* left;
    Node* right;

    Node(char character, int frequency) {
        ch = character;
        freq = frequency;
        left = nullptr;
        right = nullptr;
    }

    Node(int frequency, Node* l, Node* r) {
        ch = '\0';
        freq = frequency;
        left = l;
        right = r;
    }
};

// Comparator for Min Heap
struct Compare {
    bool operator()(Node* a, Node* b) {
        return a->freq > b->freq;
    }
};

// Build Frequency Table
unordered_map<char, int> buildFrequencyTable(const string& text) {
    unordered_map<char, int> freqTable;

    for (char c : text) {
        freqTable[c]++;
    }

    return freqTable;
}

// Build Huffman Tree
Node* buildHuffmanTree(const unordered_map<char, int>& freqTable) {
    priority_queue<Node*, vector<Node*>, Compare> minHeap;

    for (auto it = freqTable.begin(); it != freqTable.end(); ++it) {
        minHeap.push(new Node(it->first, it->second));
    }

    if (minHeap.empty())
        return nullptr;

    if (minHeap.size() == 1)
        return minHeap.top();

    while (minHeap.size() > 1) {
        Node* left = minHeap.top();
        minHeap.pop();

        Node* right = minHeap.top();
        minHeap.pop();

        Node* parent =
            new Node(left->freq + right->freq, left, right);

        minHeap.push(parent);
    }

    return minHeap.top();
}

// Generate Huffman Codes
void buildCodes(
    Node* root,
    string code,
    unordered_map<char, string>& codes) {

    if (!root)
        return;

    if (!root->left && !root->right) {
        if (code.empty())
            code = "0";

        codes[root->ch] = code;
        return;
    }

    buildCodes(root->left, code + "0", codes);
    buildCodes(root->right, code + "1", codes);
}

// Encode Text
string encode(
    const string& text,
    const unordered_map<char, string>& codes) {

    string encoded;

    for (char c : text) {
        encoded += codes.at(c);
    }

    return encoded;
}

// Decode Text
string decode(const string& encoded, Node* root) {

    if (!root)
        return "";

    if (!root->left && !root->right) {
        return string(encoded.size(), root->ch);
    }

    string result;
    Node* current = root;

    for (char bit : encoded) {

        if (bit == '0')
            current = current->left;
        else
            current = current->right;

        if (!current->left && !current->right) {
            result += current->ch;
            current = root;
        }
    }

    return result;
}

// Compression Ratio
double compressionRatio(
    const string& original,
    const string& encoded) {

    if (original.empty())
        return 0.0;

    double originalBits =
        static_cast<double>(original.size()) * 8.0;

    double compressedBits =
        static_cast<double>(encoded.size());

    return (originalBits - compressedBits) / originalBits;
}

// Full Compression Process
tuple<string, Node*, unordered_map<char, string>, double>
huffmanCompress(const string& text) {

    auto freqTable = buildFrequencyTable(text);

    Node* tree = buildHuffmanTree(freqTable);

    unordered_map<char, string> codes;

    buildCodes(tree, "", codes);

    string encoded = encode(text, codes);

    double ratio =
        compressionRatio(text, encoded);

    return make_tuple(
        encoded,
        tree,
        codes,
        ratio
    );
}

// Delete Tree
void deleteTree(Node* root) {

    if (!root)
        return;

    deleteTree(root->left);
    deleteTree(root->right);

    delete root;
}

// Main Function
int main() {

    string text;

    cout << "Enter text: ";
    getline(cin, text);

    if (text.empty()) {
        cout << "Empty string!" << endl;
        return 0;
    }

    auto result = huffmanCompress(text);

    string encoded = get<0>(result);
    Node* tree = get<1>(result);
    unordered_map<char, string> codes = get<2>(result);
    double ratio = get<3>(result);

    cout << "\n===== HUFFMAN CODES =====\n";

    for (auto it = codes.begin(); it != codes.end(); ++it) {

        if (it->first == ' ')
            cout << "' ' -> ";
        else
            cout << it->first << " -> ";

        cout << it->second << endl;
    }

    cout << "\n===== RESULT =====\n";

    cout << "Original text:\n";
    cout << text << endl;

    cout << "\nEncoded string:\n";
    cout << encoded << endl;

    cout << "\nDecoded string:\n";
    cout << decode(encoded, tree) << endl;

    cout << "\nOriginal size: "
        << text.size() * 8
        << " bits" << endl;

    cout << "Compressed size: "
        << encoded.size()
        << " bits" << endl;

    cout << "Compression ratio: "
        << ratio * 100
        << "%" << endl;

    deleteTree(tree);

    return 0;
}

File: huffman.py
import heapq
from collections import Counter

class HuffmanNode:
    """Узел дерева Хаффмана."""
    def __init__(self, char, freq):
        self.char = char          # символ (для листьев), для внутренних узлов None
        self.freq = freq
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.freq < other.freq

def build_frequency_table(text):
    """Подсчёт частот символов."""
    return Counter(text)

def build_huffman_tree(freq_table):
    """Построение дерева Хаффмана."""
    if not freq_table:
        return None

    heap = [HuffmanNode(ch, f) for ch, f in freq_table.items()]
    heapq.heapify(heap)

    # Особый случай: только один символ
    if len(heap) == 1:
        return heap[0]

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)
        merged = HuffmanNode(None, left.freq + right.freq)
        merged.left = left
        merged.right = right
        heapq.heappush(heap, merged)

    return heap[0]

def build_codes(node, prefix="", codes=None):
    """Рекурсивный обход дерева для генерации кодов."""
    if codes is None:
        codes = {}
    if node is None:
        return codes
    # Лист
    if node.char is not None:
        # Если дерево состоит из одного узла, кодом будет "0"
        codes[node.char] = prefix if prefix else "0"
    else:
        build_codes(node.left, prefix + "0", codes)
        build_codes(node.right, prefix + "1", codes)
    return codes

def encode(text, codes):
    """Кодирование строки в битовую последовательность (строка из '0'/'1')."""
    if not text:
        return ""
    return "".join(codes[ch] for ch in text)

def decode(encoded, root):
    """Декодирование битовой последовательности."""
    if not encoded or root is None:
        return ""
    # Особый случай: дерево из одного листа (все символы одинаковы)
    if root.left is None and root.right is None:
        # Каждый бит игнорируется, возвращаем символ столько раз, сколько битов
        return root.char * len(encoded)

    result = []
    node = root
    for bit in encoded:
        node = node.left if bit == "0" else node.right
        if node.char is not None:   # достигли листа
            result.append(node.char)
            node = root
    return "".join(result)

def compression_ratio(original, encoded):
    """Коэффициент сжатия."""
    if not original:
        return 0.0
    return 1 - len(encoded) / (len(original) * 8)

def huffman_compress(text):
    """Полный цикл сжатия: возвращает (закодированная_строка, дерево, коды, коэффициент)."""
    if not text:
        return "", None, {}, 0.0

    freq = build_frequency_table(text)
    tree = build_huffman_tree(freq)
    codes = build_codes(tree)
    encoded = encode(text, codes)
    ratio = compression_ratio(text, encoded)
    return encoded, tree, codes, ratio

